packages/core/saved-objects/core-saved-objects-migration-server-internal/src/zdt/next.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { AllActionStates, InitState, State } from './state';
+import type { MigratorContext } from './context';
+import * as Actions from './actions';
+
+export type ActionMap = ReturnType<typeof nextActionMap>;
+
+/**
+ * The response type of the provided control state's action.
+ *
+ * E.g. given 'INIT', provides the response type of the action triggered by
+ * `next` in the 'INIT' control state.
+ */
+export type ResponseType<ControlState extends AllActionStates> = Awaited<
+  ReturnType<ReturnType<ActionMap[ControlState]>>
+>;
+
+export const nextActionMap = (context: MigratorContext) => {
+  return {
+    INIT: (state: InitState) =>
+      Actions.init({ client: context.elasticsearchClient, indices: [context.indexPrefix] }),
+  };
+};
+
+export const next = (context: MigratorContext) => {
+  const map = nextActionMap(context);
+
+  return (state: State) => {
+    const delay = <F extends (...args: any) => any>(fn: F): (() => ReturnType<F>) => {
+      return () => {
+        return state.retryDelay > 0
```

Comment:
Also encountered an issue trying to factorize this. `nextActionMap` cannot be factorized, and `next` depends on it (and most importantly, on the shape of the `State`).

We could try spending some time to find a proper TS way to make this generic, but for 50 lines, it didn't felt like a priority.